### ISSUE TITLE
PodManager Statemachine

### ISSUE THREAD

**Issue by brightcoder01** [sql-machine-learning/elasticdl#2395]
## What
Apply the state machine design in the pod state management in `PodManager` in order to avoid unnecessary PodEventCallback invoke and improve the code maintainability.

## Why
At the current implementation, the core logic of maintaining the pods of elasticdl cluster is in `PodManager._event_cb`. The method will react with each event fired from Kubernetes and take action according to the `event_type` and `phase` property in the event object.
For example, we will refresh the Rendezvous service for the event with type `ADDED | MODIFIED` and phase `RUNNING`; and we will relaunch the worker if the event (`MODIFIED` type and `FAILED` phase or `DELETED` type).

But we will take actions too frequently and many of them are unnecessary. Please check the following scenario.
- Launch the ElasticDL Job with 1 PS and 2 Workers
- `kubectl delete pod elasticdl-test-mnist-worker-0`
- `kubectl delete pod elasticdl-test-mnist-worker-1`

```TXT
[2020-12-01 03:25:51,377] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-master   , ip:172.17.0.3  , phase:Running .
[2020-12-01 03:25:51,384] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-master   , ip:172.17.0.3  , phase:Running .
[2020-12-01 03:25:56,410] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Running .
[2020-12-01 03:25:56,414] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-master   , ip:172.17.0.3  , phase:Running .
[2020-12-01 03:25:56,417] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-0 , ip:172.17.0.5  , phase:Running .
[2020-12-01 03:25:56,420] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-ps-0     , ip:172.17.0.4  , phase:Running .
[2020-12-01 03:26:10,631] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-worker-0 , ip:172.17.0.5  , phase:Running .
[2020-12-01 03:26:14,027] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-worker-0 , ip:172.17.0.5  , phase:Failed  .
[2020-12-01 03:26:19,114] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-2 , ip:172.17.0.5  , phase:Running .
[2020-12-01 03:26:19,119] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-master   , ip:172.17.0.3  , phase:Running .
[2020-12-01 03:26:19,122] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-0 , ip:172.17.0.5  , phase:Failed  .
[2020-12-01 03:26:19,144] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Running .
[2020-12-01 03:26:19,160] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-ps-0     , ip:172.17.0.4  , phase:Running .
[2020-12-01 03:26:23,780] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-worker-0 , ip:172.17.0.5  , phase:Failed  .
[2020-12-01 03:26:23,806] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:DELETED , name:elasticdl-test-mnist-worker-0 , ip:172.17.0.5  , phase:Failed  .
[2020-12-01 03:27:21,071] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Running .
[2020-12-01 03:27:24,097] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Failed  .
[2020-12-01 03:27:29,196] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Failed  .
[2020-12-01 03:27:29,214] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-2 , ip:172.17.0.5  , phase:Running .
[2020-12-01 03:27:29,231] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-master   , ip:172.17.0.3  , phase:Running .
[2020-12-01 03:27:29,253] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-ps-0     , ip:172.17.0.4  , phase:Running .
[2020-12-01 03:27:29,295] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:ADDED   , name:elasticdl-test-mnist-worker-3 , ip:172.17.0.6  , phase:Running .
[2020-12-01 03:27:33,763] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:MODIFIED, name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Failed  .
[2020-12-01 03:27:33,800] [INFO] [pod_manager.py:443:_event_cb] Kubernetes Event. type:DELETED , name:elasticdl-test-mnist-worker-1 , ip:172.17.0.6  , phase:Failed  .
```

We can see that if we delete `elasticdl-test-mnist-worker-0`, the ADDED event for master / ps / worker-1 will also be fired. If it's a AllReduce training, the Rendezvous service will be refresh multiple times.
And also two event with `MODIFIED` type and `FAILED` phase and one event with `DELETED` type and `FAILED` phase are fired for worker-0. And we use the following code and operate the internal dict to avoid the multiple relaunch actions, but it's not easy for code maintainability. 
https://github.com/sql-machine-learning/elasticdl/blob/08ec7791f0d17a4d21af2eb983cd969526755b70/elasticdl/python/master/pod_manager.py#L473-L481

## How
![pod_manager_state_machine](https://user-images.githubusercontent.com/55301748/101018370-62997700-35a6-11eb-9902-2a13080c11eb.jpg)

Design a state machine for the pod management. Only the kubernetes event with specific type and phase will trigger the state change. And the event marked with green will trigger the pod relaunch.

**Comment by skydoorkai** [comment]
Looks good.

Also, we need to consider how to support vpa later.

**Comment by workingloong** [comment]
Rendezvous service will be refreshed multiple times but the rendezvous will not be re-initialized if the worker hosts do not change.
https://github.com/sql-machine-learning/elasticdl/blob/08ec7791f0d17a4d21af2eb983cd969526755b70/elasticdl/python/master/rendezvous_server.py#L50-L59

**Comment by brightcoder01** [comment]
The state machine is implemented in the PR #2399
